### ISSUE TITLE
[BUG] Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException

### ISSUE THREAD

**Issue by Maia-Everett** [projectlombok/lombok#2999]
**Describe the bug**
A particular line makes compilation with javac occasionally (nondeterministically) fail with an NPE in `HandleVal`. Whether or not it fails seems to depend (if it depends on anything at all) on which specific other files were modified prior to rebuilding the project.

**To Reproduce**

Here's the class with the offending line:

```java
import lombok.RequiredArgsConstructor;
import lombok.val;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.google.common.collect.Maps;

public class DependentEntityListMerger {
	@RequiredArgsConstructor
	public static abstract class Descriptor<ME extends IEntity, DE extends IEntity, LE extends IEntity> {
		private final String notFoundErrorMessage;
		private final boolean idLinksToSelf;

		public abstract Integer getLinkedId(DE entity);
		public abstract DE create(ME header, LE linkedEntity);
		public abstract List<LE> findByIds(List<Integer> ids);
	}

	public static <ME extends IEntity, DE extends IEntity, LE extends IEntity> void mergeList(
			ME header,
			List<DE> dependentEntityList,
			List<Integer> dtoIdList,
			List<String> errors,
			Descriptor<ME, DE, LE> descriptor) {
		val existingEntitiesByLinkedId = Maps.uniqueIndex(dependentEntityList, descriptor::getLinkedId);
		// ...rest removed
	}
}
```

The `IEntity` interface is declared in another package as:

```java
public interface IEntity {
	Integer getId();
}
```

**Expected behavior**
The compilation succeeds and the `val` is type-inferred as `com.google.common.collect.ImmutableMap<Integer, DE>`.

**Version info:**
 - Lombok 1.18.20
 - javac 1.8.0_292
 - Guava 19.0 (for `com.google.common.collect`)

**Additional context**
I can attach the full source if required.

Stack trace:

```
Exception while resolving: NODE LOCAL (class com.sun.tools.javac.tree.JCTree$JCVariableDecl) @val()
final ___Lombok_VAL_Attrib__ existingEntitiesByLinkedId = Maps.uniqueIndex(dependentEntityList, descriptor::getLinkedId)(RegularFileObject[/<DATA EXPUNGED>/DependentEntityListMerger.java])
java.lang.NullPointerException
	at com.sun.tools.javac.comp.Check$Validator.visitTypeApply(Check.java:1259)
	at com.sun.tools.javac.tree.JCTree$JCTypeApply.accept(JCTree.java:2135)
	at com.sun.tools.javac.comp.Check$Validator.validateTree(Check.java:1359)
	at com.sun.tools.javac.comp.Check.validate(Check.java:1230)
	at com.sun.tools.javac.comp.Attr.visitVarDef(Attr.java:1055)
	at com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:852)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
	at com.sun.tools.javac.comp.Attr.attribStats(Attr.java:661)
	at com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1124)
	at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:909)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribStat(Attr.java:645)
	at lombok.javac.JavacResolution.attrib(JavacResolution.java:249)
	at lombok.javac.JavacResolution.memberEnterAndAttribute(JavacResolution.java:217)
	at lombok.javac.JavacResolution.resolveMethodMember(JavacResolution.java:165)
	at lombok.javac.handlers.HandleVal.endVisitLocal(HandleVal.java:144)
	at lombok.javac.JavacNode.traverse(JavacNode.java:122)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
	at lombok.javac.JavacNode.traverse(JavacNode.java:105)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
	at lombok.javac.JavacNode.traverse(JavacNode.java:95)
	at lombok.javac.JavacAST.traverseChildren(JavacAST.java:222)
	at lombok.javac.JavacNode.traverse(JavacNode.java:90)
	at lombok.javac.JavacAST.traverse(JavacAST.java:218)
	at lombok.javac.HandlerLibrary.callASTVisitors(HandlerLibrary.java:276)
	at lombok.javac.JavacTransformer.transform(JavacTransformer.java:64)
	at lombok.javac.apt.LombokProcessor.process(LombokProcessor.java:328)
	at lombok.core.AnnotationProcessor$JavacDescriptor.process(AnnotationProcessor.java:187)
	at lombok.core.AnnotationProcessor.process(AnnotationProcessor.java:241)
	at lombok.launch.AnnotationProcessorHider$AnnotationProcessor.process(AnnotationProcessor.java:90)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:802)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:713)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1043)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1184)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
	at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:126)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:169)
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:785)
	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:129)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
	at org.codehaus.classworlds.Launcher.main(Launcher.java:47)
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] Lombok visitor handler class lombok.javac.handlers.HandleVal failed: java.lang.NullPointerException
[INFO] 1 error
[INFO] -------------------------------------------------------------
```

**Comment by Rawi01** [comment]
Can you try to use 1.18.22? I fixed a bug (#2420) with a slightly different stacktrace some months ago.

**Comment by rzwitserloot** [comment]
@Maia-Everett Good writeup, and 1.18.22 was released rather recently :) I'll close this but if you run into it again, leave a comment and I'll re-open.
